tools/chpl-language-server/test/type-inlays.py

```diff
@@ -0,0 +1,491 @@
+            sys.executable,
+            CLS_PATH(),
+            "--resolver",
+            "--type-inlays",
+            "--no-literal-arg-inlays",
+            "--no-param-inlays",
+            "--end-markers=none",
+        ],
+        client_factory=get_base_client,
+    )
+)
+async def client(lsp_client: LanguageClient):
+    # Setup
+    params = InitializeParams(capabilities=ClientCapabilities())
+    await lsp_client.initialize_session(params)
+
+    yield
+
+    # Teardown
+    await lsp_client.shutdown_session()
+
+
+@pytest.mark.asyncio
+async def test_type_inlays_prim(client: LanguageClient):
+    """
+    Ensure that type inlays are shown for primitive types.
+    """
+
+    file = """
+            var a = 10;
+            var b: int = 10;
+            var c = a:int(32);
+            var d = 0.0;
+
+            var e = "hello";
+            var f = true;
+           """
+
+    inlays = [
+        (pos((0, 5)), "int(64)"),
+        (pos((2, 5)), "int(32)"),
+        (pos((3, 5)), "real(64)"),
+        (pos((5, 5)), "string"),
+        (pos((6, 5)), "bool"),
+    ]
+
+    with source_file(client, file) as doc:
+        await check_type_inlay_hints(
+            client, doc, rng((0, 0), endpos(file)), inlays
+        )
+
+        # check that specifying a range works
+        await check_type_inlay_hints(
+            client, doc, rng((0, 0), (3, 0)), inlays[:2]
+        )
+
+        await save_file(client, doc)
+        assert len(client.diagnostics[doc.uri]) == 0
+
+
+@pytest.mark.asyncio
+@pytest.mark.xfail
+async def test_type_inlays_user_defined_types(client: LanguageClient):
+    """
+    Ensure that type inlays are shown properly for user-defined types.
+    """
+    file = """
+            {at} Concrete {{
+              var x: int;
+            }}
+            {at} Generic {{
+              type t;
+              var x: t;
+              proc init(type t) {{
+                this.t = t;
+              }}
+            }}
+            var a = new Concrete();
+            var b = a;
+
+            var c = new Generic(int);
+            var d = c;
+
+            var e = new Generic(Concrete);
+            var f = e;
+
+            var g = new Generic(Generic(string));
+            var h = g;
+           """
+
+    inlay_pos = [
+        (pos((10, 5)), "{m}Concrete"),
+        (pos((11, 5)), "{m}Concrete"),
+        (pos((13, 5)), "{m}Generic(int)"),
+        (pos((14, 5)), "{m}Generic(int)"),
+        (pos((16, 5)), "{m}Generic({m}Concrete)"),
+        (pos((17, 5)), "{m}Generic({m}Concrete)"),
+        (pos((19, 5)), "{m}Generic({m}Generic(string))"),
+        (pos((20, 5)), "{m}Generic({m}Generic(string))"),
+    ]
+
+    variants = [("record", ""), ("class", "owned ")]
+
+    for at, management in variants:
+        with source_file(client, file.format(at=at)) as doc:
+            inlays = [
+                (pos, typename.format(m=management))
+                for pos, typename in inlay_pos
+            ]
```

Comment:
Wow, I just added this function to a C++ test I was writing. It was even called 'versions', which starts with the same letter! Great minds?